### PR TITLE
fix: apply reward_weights to logged reward/reward_std in GRPOTrainer

### LINKED ISSUE #5352: GRPOTrainer: logged reward/reward_std metrics ignore reward_weights
### System info

```
TRL main branch (commit on current HEAD)
Python 3.13, PyTorch 2.x, transformers latest
```

### Bug description

In `GRPOTrainer._generate_and_score_completions()`, the `reward` and `reward_std` metrics logged to training trackers (WandB, MLflow, TensorBoard) use an **unweighted** sum of per-function rewards, even when `reward_weights` are configured with non-uniform values.

The advantage computation correctly applies `self.reward_weights` (lines 1965 and 1997 in `grpo_trainer.py`), so models train correctly. However, the logging section at line 2022 overwrites the weighted `rewards` variable with `rewards_per_func.nansum(dim=1)` (unweighted) before logging:

```python
# Line 2022 — overwrites the correctly-weighted rewards variable:
rewards = rewards_per_func.nansum(dim=1)
self._metrics[mode]["reward"].append(rewards.mean().item())
self._metrics[mode]["reward_std"].append(rewards.std().item())
```

This means the dashboard `reward` metric diverges from the actual training signal whenever `reward_weights` are non-uniform.

### Reproduction

```python
# Two reward functions: one always returns 1.0, the other always returns 0.0
# With reward_weights=[0.7, 0.3]:
#   Expected logged reward: 0.7*1.0 + 0.3*0.0 = 0.7
#   Actual logged reward:   1.0 + 0.0 = 1.0

from trl import GRPOTrainer, GRPOConfig

config = GRPOConfig(
    output_dir="test_output",
    reward_weights=[0.7, 0.3],
    num_generations=4,
    report_to="none",
)

def reward_a(completions, **kwargs):
    return [1.0] * len(completions)

def reward_b(completions, **kwargs):
    return [0.0] * len(completions)

trainer = GRPOTrainer(
    model="Qwen/Qwen2.5-0.5B-Instruct",
    args=config,
    reward_funcs=[reward_a, reward_b],
    train_dataset=dataset,
)
trainer.train()
# trainer.state.log_history[-1]["reward"] == 1.0, should be 0.7
```

### Expected behavior

The logged `reward` metric should equal `mean(weighted_sum_per_sample)`, matching the actual training signal. With default weights (all 1.0), behavior is unchanged since weighted sum equals unweighted sum.

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/huggingface/trl/issues) in the repo

### PR DESCRIPTION
**Pull request by lailanelkoussy**
## Summary
- The `reward` and `reward_std` metrics logged by `GRPOTrainer` use `rewards_per_func.nansum(dim=1)` (unweighted), even when `reward_weights` are non-uniform
- The actual training signal (advantages) correctly applies `reward_weights`, so this is a logging-only bug
- This PR applies `self.reward_weights` before the nansum in the logging section, making the dashboard metrics match the optimization objective

Fixes #5352

## Changes
- `trl/trainer/grpo_trainer.py`: Replace unweighted `nansum` with weighted aggregation for metric logging (3 lines)
- `tests/test_grpo_trainer.py`: Add test verifying logged reward equals weighted sum with non-uniform weights

## Notes
- Default `reward_weights=[1.0, ...]` → no behavior change (weighted sum = unweighted sum)
- Both `sum_then_normalize` and `normalize_then_sum` paths are covered (logging runs after the if/elif)
- Per-function metrics (`rewards/func_name/mean`) are unchanged (they are per-function, not aggregated)

---

> [!NOTE]
> **Low Risk**
> Low risk: changes only how `reward`/`reward_std` are aggregated for logging/metrics, but could affect dashboards and any automation that keys off these logged values when non-uniform `reward_weights` are used.
> 
> **Overview**
> Fixes aggregated reward metric logging to **apply `reward_weights`** when computing `reward` and `reward_std`, aligning reported metrics with the weighted multi-reward objective.
> 
> This updates GRPO/RLOO trainers (and related experimental trainers) to weight `rewards_per_func` before summing, updates docs to describe the weighted aggregation, and adds regression tests ensuring logged `reward` reflects non-uniform weights (including both GRPO multi-objective aggregation modes).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 14db5527d27d1dc8b6b3c4bce5aa74aee5eba671. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -2019,9 +2019,9 @@ def _generate_and_score_completions(
             self._metrics[mode][f"rewards/{reward_func_name}/mean"].append(mean_rewards)
             std_func_rewards = nanstd(rewards_per_func[:, i]).item()
             self._metrics[mode][f"rewards/{reward_func_name}/std"].append(std_func_rewards)
-        rewards = rewards_per_func.nansum(dim=1)
-        self._metrics[mode]["reward"].append(rewards.mean().item())
-        self._metrics[mode]["reward_std"].append(rewards.std().item())
+        weighted_rewards = (rewards_per_func * self.reward_weights.to(device).unsqueeze(0)).nansum(dim=1)
+        self._metrics[mode]["reward"].append(weighted_rewards.mean().item())
+        self._metrics[mode]["reward_std"].append(weighted_rewards.std().item())
```
[the remark this answers is not in the document]
Comment (lailanelkoussy) [reply trl/trainer/grpo_trainer.py]: We intentionally recompute from `rewards_per_func` (raw values) rather than reusing the already-computed `rewards` variable, because for `normalize_then_sum` mode, `rewards` holds the weighted sum of *normalized* rewards, while per-function metrics (`rewards/func_name/mean`) log raw values. Using raw weighted sums keeps the aggregate `reward` metric consistent with the per-function breakdown.

Added a `normalize_then_sum` test variant to cover this.

**Comment by qgallouedec** [comment]
Thanks. We originally implemented it that way on purpose, but after thinking it over, I think you're right, it makes sense for the weighting to affect the total logged reward. It feels more natural and intuitive.

I've updated your branch so that the same change is applied in RLOO and elsewhere in the library, as well as in the documentation.

**Comment by qgallouedec** [comment]
@codex review

File: trl/experimental/grpo_with_replay_buffer/grpo_with_replay_buffer_trainer.py [hunk 2]
```diff
@@ -332,8 +332,9 @@ def _generate_and_score_completions(
             self._metrics[mode][f"rewards/{reward_func_name}/mean"].append(mean_rewards)
             std_func_rewards = nanstd(rewards_per_func[:, i]).item()
             self._metrics[mode][f"rewards/{reward_func_name}/std"].append(std_func_rewards)
-        self._metrics[mode]["reward"].append(mean_grouped_rewards.mean().item())
-        self._metrics[mode]["reward_std"].append(std_rewards.mean().item())
+        rewards = (rewards_per_func * self.reward_weights.to(rewards_per_func.device).unsqueeze(0)).nansum(dim=1)
+        self._metrics[mode]["reward"].append(rewards.mean().item())
+        self._metrics[mode]["reward_std"].append(rewards.std().item())
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/grpo_with_replay_buffer/grpo_with_replay_buffer_trainer.py:337]: out of the scope

[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/grpo_with_replay_buffer/grpo_with_replay_buffer_trainer.py:337]: actually, not out of the scope, fixing

**Review by qgallouedec — APPROVED** [review]
